### ISSUE TITLE
String.assign performance

### ISSUE THREAD

**Issue by ermouth** [andrewplummer/Sugar#535]
I‘ve discovered that `String.assign` performance is subotimal for complex templates. 

``` javascript
var s = '{x}  {y} '+('{yy}').repeat(1000)+' {x} {z}',
a = {x:"123",y:[5,6],z:3,yy:1},
t = Date.now(); 
for(var i=0;i<10000;i++) s.assign(a);
console.log(Date.now()-t);
```

runs about 4 seconds.

If we change an approach from in-place handling of each `{smth}` match to pre-generating regexps for each key in data obj, we can gain about 10x performance boost for an example given. Code

``` javascript
function assign(str, obj){
    var keys = Object.keys(obj), i=0, acc;
    if (keys.length>100) return str.assign(obj);
    for (acc=str;i<keys.length;i++) 
        acc = acc.replace(new RegExp('\\{'+RegExp.escape(keys[i])+'\\}','g'), obj[keys[i]]);
    return acc;
}
t = Date.now(); 
for(var i=0;i<10000;i++) assign(s,a);
console.log(Date.now()-t);
```

runs about half a second.

Although for small templates this approach is bit slower.

**Comment by andrewplummer** [comment]
First, `String#assign` will be `String#format` as per #476. The underlying implementation has changed as well, so If you can please check out `master-2.0.0` branch and compare performance to that, it would be helpful.

Also I should have mentioned this a while ago (as I know you're a performance nut ;) ) but there's now some baked in performance tests as well. After checking out the above branch open `test/browser/performance/index.html`. You can tweak that file and `test.js`. These aren't meant to be up to date, or continuous or anything, just basically a poor-man's jsperf.com (and already set up for Sugar). Anyway should be useful for you. You can just copy/paste your `assign` implementation in there and run it as test 2.

Last, I realize you're probably more interested in this for 1.5.0 (which will be the next v1) but if we can prove that this really is much better for performance, then I can have a look at retrofitting it back to v1.

**Comment by vendethiel** [comment]
It's nice to know you have perf tests!

**Comment by ermouth** [comment]
@andrewplummer Well, I’ve tested `String.format` of `master-2.0.0`. Actually, 2.0 performance is ~3 times worse :(

So as for 2.0 implementation, my example outperformes it at least twice even for very short templates, and 20-30 times for templates with thousands of entries.

**Comment by andrewplummer** [comment]
Ok, so a few comments here. First, I've added perf tests (linked above) for String#format, and it seems that you're right about it being slower, but only for the test case you've outlined. I don't see any performance change for short templates.

Anyway, using the new `String#format` as a base, you are correct that for "complex" template strings, it is about 2x slower than your implementation, however for short template strings it is 2x faster (as you say in your first comment).

However, your test case is also assuming repeating keys. If we use unique keys instead, your implementation appears to be 2x slower for not only short template strings but also equivalently "complex" strings, but with unique keys. It seems that the performance gain here only happens when keys are used multiple times in the template.

I've committed test cases for all of these so you can check.

Additionally for completeness, I've also added a version that doesn't use regexes at all but instead walks through the string character by character.

**Comment by ermouth** [comment]
Well, I can‘t give up so early :) This is equal or bit faster for shortest templates in 1.4.1, regardless of obj size, and about 1/3 faster for long templates with repeating entries.

``` javascript
function assign(str, obj){
    var keys = Object.keys(obj), i=0, acc=[],
    a = str.split("{"), pos=0, key="", v, r;
    if (a.length==1) return str;
    for (;i<a.length;i++){
    	if (a[i].length) {
	    	if ((pos=a[i].indexOf("}"))>0) {
	    		key = a[i].substr(0,pos);
				v=obj[key];
	    		if (null != v) acc.push(v+a[i].substr(pos+1));
	    		else acc.push("{",a[i])
	    	} else acc.push("{",a[i]);
	    }
    }
    return acc.join("");
}
t = Date.now(); 
for(var i=0;i<10000;i++) assign(s,a);
console.log(Date.now()-t);
```

**Comment by andrewplummer** [comment]
Nope, I don't want to give up either! 

So I added your new case...splitting up front is an interesting idea, and I also think `indexOf` could be more cleverly applied as well.

Your latest version is performing well, however I forgot this earlier, but the new `format` method allows deep keys (I think you were maybe involved in that thread as well) so it's not as simple as checking for keys in the object anymore, so I don't think this will work. However your use of splitting and indexOf may work.

Also, I mentioned my non-regex character-walking version above but didn't really push it because I thought it was slower for long templates. However I had made a mistake, and it turns out that is the best performing so far, except for many repeated names, in which case your first implementation is about 10x faster (strangely your latest version is 2x slower).

Anyway I'll have another look at this later. I think with a bit of massaging we can definitely squeeze more cycles out of this method. Not 100% sure about the `keys` approach for the above reasons however.

**Comment by vendethiel** [comment]
Anither way might be to precompile the templates?

**Comment by andrewplummer** [comment]
@vendethiel That's interesting although it's getting into the territory of a proper templating system like handlebars, etc... Compiling the templates up front would probably have its own performance hit. For those libs it makes sense as an HTML template is something that's static and is going to be constantly hammered with varying data. Sugar on the other hand is meant to be more one-off and for "templating" akin to ES6 string templates instead.

**Comment by vendethiel** [comment]
It could make sense with a flag, I guess, but if you want to support both, that means code duplication...

**Comment by andrewplummer** [comment]
Ok, I had a few more bashes at it... Notably:

1. Tried your method of iterating over the object keys up front before doing the standard global `{...}` replace after. The idea here is that if we replace known keys up front, then the `{}` tokens will have been removed for the more complex regex. This one has results that are more or less the same as your original improvement: 20x speed gain for your repeated keys test, but 2x slower for unique keys.

2. Tried an extremely simple version that only uses `String#indexOf` and `String#slice`. This showed gains of 2x for unique keys and 20% gain for your repeated keys test, so better than what's in place now, but not as good as yours for repeated keys.

The problem I have with your test is that the performance gains are relative to the number of times the key is repeated in the string. Even generally speaking, I would say that having all unique keys in the format string (for example `Hello, {user.firstName} {user.lastName}`) is a more common use case than having repeated keys (ex `Hi, {name}, your name is {name}`). However, your test repeats the same key 1000 times, so the gains are pretty meaningless (1000 repeated keys produces 20x gain, 2000 repeated keys produces 40x, but neither of these is realistic). If this could be done in a way that doesn't detriment unique key performance I might go for it, but so far haven't been able to do that.

Also, as mentioned above, the approach itself isn't ideal as keys may be deep. My performance tests aren't even taking that into account for simplicity's sake, but in a real world situation (for example the `{user.firstName}` above) the benefits of doing your approach up front wouldn't matter because `user.firstName` isn't a key, so it would have to fall through and hit the 2nd general regex.

Anyway, I'm going to see how far I can take the `IndexOf`/`slice` approach. So far it's showing all around better performance, so it may be a winner, but there's actually a lot of edge cases with nested `{}` tokens with this method, so we'll see after the dust settles.

**Comment by ermouth** [comment]
> I'm going to see how far I can take the IndexOf/slice approach

As for me, for this particular case the most lowlevel solutions wins. Since the case is well determined, any highlevel solution will finally do steps lowlevel solution does, but with additional overhead for function calls, regexp finite automaton compilation/application, may be unnecessary memcpy and so on.

**Comment by andrewplummer** [comment]
> As for me, for this particular case the most lowlevel solution wins.

I thought about this statement for a while, thought I agreed with it, then changed my mind a bit. In theory you're right, but this doesn't take into account the fact that pre-existing solutions to well determined problems (such as regexes themselves), can (again in theory) take advantage of internal optimizations that reach even lower level than the lowest level JS. Or to put it another way, there are times when rolling your own low-level solution can actually be slower. It's hard to think of good examples of this but I know they exist, as I have often been surprised by higher level methods that would seem to have overhead outperform a custom solution. Of course this isn't always the case, but often it's for other reasons, such as `forEach` performing various checks to comply with spec, making it  notoriously slower than a for loop.

Additionally, while performance is very important, I would say that the benefit of concise, easy to understand code is non-zero, especially when you consider both filesize overheads and also edge cases that custom low level solutions can overlook.

Anyway... sorry, went off on a tangent...

But that leads us here. I've tried about 8 different approaches now (all published) and taken 2 all the way -- meaning that I've gotten all the unit tests to pass. For this method that's important because there are a lot of edge cases. So far, walking the characters is the fastest approach, but after handling all the edge cases, the performance benefits are not super exciting, and the resulting code is pretty  ugly. Compared to other operations in Sugar, I don't consider this method to be a bottleneck in the first place, so I'm not extremely motivated to make this change.

In fact, the more I think about it, if we're going to have actual, meaningful impact on end-user performance then @vendethiel's compiling solution of compiling templates makes even more sense than any minute code changes. I actually forgot that I'm already doing this with the `Date#format` method anyway. It may even be possible to consolidate the two.

Anyway I think I'm going to look at this from that approach instead. However you're welcome to look at what I've done and try out a better solution. The performance tests are easy to run, however the only caveat is that you have to make the unit tests pass too, with one exception: the ones that assert that empty `{}` must be unmatched and produce `{}` can be ignored and return `undefined` instead as an edge case.

**Comment by andrewplummer** [comment]
@vendethiel Well this is interesting. Trying out caching (i.e. precompiling) the format string. The results are better but they're nothing like the gains I was expecting. Feel free to check my work... I've double checked, and the formats are being cached into a simple array of functions, but the gains are only along the lines of 10-20%.

This is leading me to believe that we really are pretty damn close to the metal here. The fact that the overhead of making 1000 function calls is only marginally better than parsing out 1000 tokens with a regex, and somewhat comparable with walking the string by character tells me that this method is already at a point of diminishing returns with optimizations.

I would like to go with this solution anyway as 1. it is faster, if only marginally, 2. It dovetails nicely with `Date#format` which can reuse the same code, 3. I've made some other performance enhancements that should benefit `Date#format` as well. However, I think I'm close to calling this one closed.

**Comment by vendethiel** [comment]
I can't really look now... It has to be done carefully, because eval/function.new can "poison" the outer scope and prevent some optimizations

**Comment by andrewplummer** [comment]
So, I guess it depends on your definition of "precompile" but this is not performing anything like that. What I'm doing here is simply turning the format string into an array of functions that are executed in order, so that the string doesn't have to be parsed out again. This has shown some big performance gains for `Date#format`. 

What you are talking about is "function inlining" or "function templating" which uses `eval` or `new Function` and kicks off the JS parser. I've had a very long look at that in the past and long story short it's something that I've deemed as off-limits for Sugar. The performance gains from this method were less than overwhelming from what I've seen and it raises a number of security issues in production servers with Content Security Policies that will end up being confusing and simply aren't worth the risk/complexity.

**Comment by vendethiel** [comment]
how does eval relate to CSP? I don't remember anything about this kind of issue

**Comment by vendethiel** [comment]
I'd call what you're doing "specializing" the template instead, but that's probably nitpicking.

**Comment by andrewplummer** [comment]
Ya you're probably right... I do hear function inlining sometimes referred to as "precompiling" but on the other hand moment.js does the same thing I'm doing above and calls it "compiling" as well.

Anyway, any use of `eval` or `new Function()` are blacklisted for sites using a CSP unless they explicitly set `"script-src 'unsafe-eval'"`. Most server scaffolds that include a CSP don't allow unsafe-eval for obvious reasons, and it's highly recommended against. Lodash I believe had (not sure about now) a template precompile task specifically to avoid this issue.

More here: https://developer.chrome.com/extensions/contentSecurityPolicy#relaxing-eval

**Comment by vendethiel** [comment]
Thanks! I was indeed thinking about lodash.

**Comment by ermouth** [comment]
Sugar could return pre-compiled template and then allow to use it – as, for example, HB does. Like `"{a.b}{c.d}".assign()` returns precompiled template function or object with methods.

**Comment by andrewplummer** [comment]
If the format is cached internally then does it matter? Would be just simpler to call the method, no?

**Comment by andrewplummer** [comment]
I was thinking to myself I wonder why handlebars does it that way... but I guess it's maybe because it's meant as a templating not a utility library anyway...

**Comment by ermouth** [comment]
Because compilation have its cost, and templates are used not only once in general. Moreover, they could be used thousands or millions of times.

To note, making good compiled template is quite challenging task, since you have to take in account possible compiler deopts. Try/catch usage in function body leads to nearly guaranteed deopt, for example. Once I‘ve even seen `with` in jQuery.tmpl, it‘s so tempting and easy, but also leads to deopt.

**Comment by andrewplummer** [comment]
> Because compilation has its cost, and templates are used not only once in general. Moreover, they could be used thousands or millions of times.

Right but that's why it would be caching them... or am I missing something?

> To note, making good compiled template is quite challenging task, since you have to take in account possible compiler deopts.

Yeah that's a very good point, and I'm trying very hard to not have Sugar be the perfect templating system as well (that should be left to more dedicated libs). On the other hand better to have those things in mind.. I've actually gone over the major deopts with Sugar as a whole recently. Will give it a double pass when re-examining this method specifically.

**Comment by andrewplummer** [comment]
Also good point about "used not only once in general". I was thinking this as well. It's definitely true for date formatting and should be the assumption with string formatting as well I think.

**Comment by ermouth** [comment]
> Right but that's why it would be caching them... or am I missing something?

How you gonna cache? There are, as I see, 3 options:
1. Use a template string as a key in cache repo
2. Use quick hash of a template string as a key in cache repo
3. Attach compiled template to the source string as an additional property.

Both 1 and 2 are not so nice, since they allocate memory and never free it, that is absolutely unsuitable for server use.

**Comment by andrewplummer** [comment]
Definitely 1. I understand that this may not be ideal, but I don't see the other solutions as being nice either. Unfortunately there are other areas where Sugar uses caching like this and I think it has tradeoffs like anything else.

**Comment by andrewplummer** [comment]
I mean "allocating memory and never freeing it" is almost the definition of caching. I realize that proper server caches have a mechanism for freeing memory of course, but this is for things like inbound requested data that piles up, not for smaller utility libs. If this is really an issue then maybe I should do something like run through all the code that uses memory caches like this and create something like `Sugar.purge` as a mechanism to delete these caches. In any case it shouldn't be something given consideration in relationship to this method alone.

**Comment by ermouth** [comment]
> "allocating memory and never freeing it" is almost the definition of caching

Of memoization, surely.

> not for smaller utility libs

Don‘t be so shy ) Your particular ‘small utility lib‘ serves more then million requests daily server side in my projects only. 

We surely have mechanics, that free frozen RAM, since nearly every lib in JS world leaks – but I‘d prefer to bother much less about it. So `Sugar.purge` is definitely good idea.

**Comment by andrewplummer** [comment]
> Of memoization, surely.

Right, definitely. I'll say "memoization" from here on. But Sugar memoizes a looot of stuff. OK not *that* much. But there are a number of places.

> Don‘t be so shy ) Your particular ‘small utility lib‘ serves more then million requests daily server side in my projects only.

Haha okok... I can always count on you to keep pushing me.

> So Sugar.purge is definitely good idea.

I'm definitely willing to do this, but maybe there's a better way? Could Sugar purge it's own caches before getting too huge? What about using WeakMap? I'm kind of interested in this method except for the fact that it's not available in earlier environments (where it needs it most!) ... any other ideas??

**Comment by vendethiel** [comment]
WeakMap is - sadly - not shimable :/.

**Comment by andrewplummer** [comment]
Well that's kind of my point. WeakMap with (what essentially is) a normal map fallback is better than no WeakMap at all?

**Comment by ermouth** [comment]
WeakMap does not take strings as keys. You must first convert string to an object, like `new String("abc")` – and conversion is not cheap.

Also, since you converted string to an object, you actually can not later access the memoized value, because `(new String("abc")) == (new String("abc"))` returns `false`.

**Comment by andrewplummer** [comment]
@ermouth Good points... I suppose you could keep a map of primitives to objects `var map = {}; map[str] = new String(str); weakMap.get(map[str])` ... or does that take us back to the beginning again? I guess holding the strings in memory may be better than whatever objects the WeakMap might hold...

I admit, it feels pretty dumb though...

**Comment by ermouth** [comment]
I‘ve tried cache + lowlevel approach, it gives +25% speedup for small templates and about 2x or even more for long, compared to just lowlevel.

I do not memoize curried functions, it‘s expensive. I memoize array of pairs (`key`, `tail`), that is source string parsed. So `["x", "-", "y","z",null,"{"]` is result of parsing `"{x}-{y}z{"`.

```javascript
var assign = (function(){
	var repo = new Object(null);
	return function assign(str, obj){
		var i=0, acc, a, pos=0, key="", h = "", v, row;
		if (!(acc = repo[str])) {
			a = str.split("{");
			if (a.length==1) return str;
			acc = repo[str] = []; 
			for (;i<a.length;i++){
				row = a[i];
				if ((pos=row.indexOf("}"))>0) {
					key = row.substr(0,pos);
					acc.push(key, row.substr(pos+1));
				} 
				else acc.push(null, (i?"{":"")+row);
			}
		}
		for (i=0;i<acc.length;i+=2) {
			key = acc[i];
			if (key !== null) {
				v = obj[key];
				if (null != v) h+=v;
				else h+="{"+key+"}";
			}
			h+=acc[i+1];
		}
		return h;
	}
})();


var t = Date.now(); 
for(var i=0;i<10000;i++) assign(s,a);
console.log(Date.now()-t);
```

**Comment by andrewplummer** [comment]
That's an interesting approach, however on my end I'm showing no difference whatsoever between that and memoizing functions. As it stands now,  your method above is one of the top performers, but it's most likely due to just the memoization. If we're using memoization then low-level vs non-low-level makes almost no difference as the parsing is already done.

**Comment by andrewplummer** [comment]
Ok, need to make a move on this one, so I'm going with a best effort approach that combines our learnings here. This consolidates `String#format` and `Date#format` into a single format matcher that uses memoization (`Date#format` was already doing this but `String#format` was not). In addition to being clearer, more concise code, the overall average performance benefits are showing around 60% speed increase for `String#format` and 8% for `Date#format`. `Date#format` also now benefits from `String#format`'s ability to throw errors on format parsing, so this really is combining the best of both.

As to the question of memoization trapping memory, I'm going to make a separate ticket for this.

**Comment by Earl-Brown** [comment]
I have an alternative that seems nearly as performant (just hitting F5 in Chrome a lot; sometimes it's faster, usually it's about 50 milliseconds slower, and sometimes it's WAY slower).

The concept is "create a bunch of regular expressions from obj; then execute the array of regular expressions".

Here is my test code:

`var s = '{x}  {y} '+('{yy}').repeat(1000)+' {x} {z}',
a = {x:"123",y:[5,6],z:3,yy:1};

var assign_1 = (function(){
    var repo = new Object(null);
    return function assign(str, obj){
        var i=0, acc, a, pos=0, key="", h = "", v, row;
        if (!(acc = repo[str])) {
            a = str.split("{");
            if (a.length==1) return str;
            acc = repo[str] = []; 
            for (;i<a.length;i++){
                row = a[i];
                if ((pos=row.indexOf("}"))>0) {
                    key = row.substr(0,pos);
                    acc.push(key, row.substr(pos+1));
                } 
                else acc.push(null, (i?"{":"")+row);
            }
        }
        for (i=0;i<acc.length;i+=2) {
            key = acc[i];
            if (key !== null) {
                v = obj[key];
                if (null != v) h+=v;
                else h+="{"+key+"}";
            }
            h+=acc[i+1];
        }
        return h;
    }
})();

function assign_2(str, obj) {
	var expressions = [];
	for (var key in obj){
		expressions.push({
			expression: new RegExp(
				["\\{", key, "\\}"].join(""), 
				"gim"
			),
			value: obj[key]
		})
	}
	
	var replaced = str;
	expressions.forEach(function(expression) {
		replaced = replaced.replace(expression.expression, expression.value);
	});
	
	return replaced;
}

var t = Date.now(); 
for(var i=0;i<10000;i++) assign_1(s,a);
console.log("1", Date.now()-t);
console.log(assign_1(s,a));

t = Date.now();
for(var i=0;i<10000;i++) assign_2(s,a);
console.log("2", Date.now()-t);
console.log(assign_2(s,a));
`

**Comment by andrewplummer** [comment]
I've added your implementation to the perf tests for this method. First, for clarity, the method is now called 'format' instead of 'assign'. I'm also wondering which version you were testing it against. Perhaps it was the live 1.4.1 version at the time? This fix has now gone out with 2.0.0 as is much more performant than `assign` previously was. Have a look at `test/browser/performance/index.html'. It's currently showing your implementation as being a few orders of magnitude slower than the current implementation (this is testing against a few difference cases).

Feel free to tweak your implementation and try different approaches, however you should know that any final tweak to this needs to also pass all the unit tests (can be found in `test/tests/string`) as well. Additionally the current implementation is essentially the same as your approach, except that it is creating an array of functions for a given template string and executing them. Once the format is compiled, execution time is significantly reduce (2nd run on) and using all functions (even for string literals) means that the compiled format doesn't have to make a distinction between tokens and literals, also reducing execution time.

Anyway, thanks for your contribution!

**Comment by Earl-Brown** [comment]
I was only testing against the tests here.  

I hadn't thought of improvements that 2.0 would have made.  I'll have to look at the code for ideas on how to make my code better in the future!

**Comment by andrewplummer** [comment]
Finally closing (with a tinge of regret) this fun issue. `format` should now be considerably faster than it was in v1. Remember, though, that it is also handling deep keys so it's not a precise 1:1 comparison. In the end the method was given internal speed optimizations and also a memoization mechanism that should be considerably more performant, but will not leak memory (see #540). This formatting system has also been consolidated with `Date#format` as well.
